stl/inc/algorithm

```diff
@@ -2900,6 +2900,121 @@ _FwdIt3 transform(_ExPo&& _Exec, const _FwdIt1 _First1, const _FwdIt1 _Last1, co
+            requires indirectly_writable<_Out, indirect_result_t<_Fn&, projected<iterator_t<_Rng>, _Pj>>>
+        _NODISCARD constexpr unary_transform_result<borrowed_iterator_t<_Rng>, _Out>
+            operator()(_Rng&& _Range, _Out _Result, _Fn _Func, _Pj _Proj = {}) const {
+            auto _Res = _Transform_unchecked(_Ubegin(_Range), _Uend(_Range),
+                                             _STD move(_Result), _Pass_fn(_Func), _Pass_fn(_Proj));
+
+            auto _First = _RANGES begin(_Range);
+            _Seek_wrapped(_First, _STD move(_Res.in));
+            return {_STD move(_First), _STD move(_Res.out)};
+        }
+
+        template <input_iterator _It1, sentinel_for<_It1> _Se1, input_iterator _It2, sentinel_for<_It2> _Se2,
+            weakly_incrementable _Out, copy_constructible _Fn, class _Pj1 = identity, class _Pj2 = identity>
+            requires indirectly_writable<_Out, indirect_result_t<_Fn&, projected<_It1, _Pj1>, projected<_It2, _Pj2>>>
+        _NODISCARD constexpr binary_transform_result<_It1, _It2, _Out> operator()(_It1 _First1, _Se1 _Last1,
+            _It2 _First2, _Se2 _Last2, _Out _Result, _Fn _Func, _Pj1 _Proj1 = {}, _Pj2 _Proj2 = {}) const {
+            _Adl_verify_range(_First1, _Last1);
+            _Adl_verify_range(_First2, _Last2);
+
+            auto _Res = _Transform_unchecked(_Get_unwrapped(_STD move(_First1)), _Get_unwrapped(_STD move(_Last1)),
+                                             _Get_unwrapped(_STD move(_First2)), _Get_unwrapped(_STD move(_Last2)),
+                                             _STD move(_Result), _Pass_fn(_Func), _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+
+            _Seek_wrapped(_First1, _STD move(_Res.in1));
+            _Seek_wrapped(_First2, _STD move(_Res.in2));
+            return {_STD move(_First1), _STD move(_First2), _STD move(_Res.out)};
+        }
+
+        template <input_range _Rng1, input_range _Rng2, weakly_incrementable _Out, copy_constructible _Fn,
+            class _Pj1 = identity, class _Pj2 = identity>
+            requires indirectly_writable<_Out, indirect_result_t<_Fn&, projected<iterator_t<_Rng1>, _Pj1>,
+                                                                       projected<iterator_t<_Rng2>, _Pj2>>>
+        _NODISCARD constexpr binary_transform_result<borrowed_iterator_t<_Rng1>, borrowed_iterator_t<_Rng2>, _Out>
+            operator()(_Rng1&& _Range1, _Rng2&& _Range2, _Out _Result, _Fn _Func, _Pj1 _Proj1 = {}, _Pj2 _Proj2 = {})
+            const {
+            auto _Res = _Transform_unchecked(_Ubegin(_Range1), _Uend(_Range1),
+                                             _Ubegin(_Range2), _Uend(_Range2),
+                                             _STD move(_Result), _Pass_fn(_Func), _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+
+            auto _First1 = _RANGES begin(_Range1);
+            auto _First2 = _RANGES begin(_Range2);

```

Comment:
Is the current convention to call/store these first, and then unwrap them for the unchecked call, instead of separately calling `_Ubegin`?

Reply:
As I understand it we want the unwrapped iterators for the call to the private method but the wrapped iterators to return the result of the operation.

So one call to `_Ubegin` and one to `begin`. That said there is no need to create `_First1``or `_First2` earlier

Reply:
We don't bother to keep a copy of the result of `begin` around for `forward_range`s since we can simply call `begin` again. For single-pass (non-`forward_range`) ranges, however, `begin` might only be callable once. `test::range::begin` enforces that it isn't called more than once but `_Unchecked_begin` mistakenly does not (fixed by #959).

Avoiding a couple of local variables probably isn't worth `if constexpr`-ing here, you should probably just unconditionally store `_First1` and `_First2` before calling `_Transform_unchecked`.

Reply:
Absolutely. Fixed in al PRs